wordpress-mobile/WordPress-iOS
# Fix crash on Notifications tab when device is offline

The app crashed when the device was offline and the suggestions code (which handles Mentions and Cross-posts) returned a `nil` result. This result was interpreted by the `SuggestionsTableView` code as "suggestions haven't been fetched yet". The code, which has logic for fetching suggestions tied to the suggestions data source (a property), then entered an infinite loop of reloading the suggestions list.

The fix here is to undo an architectural decision made long ago to tie the fetching of suggestions to the suggestions data source. I've used `didMoveToSuperview` on the suggestions view as the trigger to fetch suggestions. This method, somewhat similar to `UIViewController`'s `viewDidLoad`, is called once each time the suggestions UI is shown.

Fixes https://github.com/wordpress-mobile/WordPress-iOS/issues/15873

### To test

**Note: While testing, you might notice inconsistencies with regard to offline support for suggestions. If you come across any, please check https://github.com/wordpress-mobile/WordPress-iOS/issues/15919 to ensure they are already tracked.**

```
- [ ] Reproduce the crash on 16.7 beta
- [ ] Verify the crash is fixed on this branch
- [ ] Regression testing
```

#### Reproduce the crash on 16.7 beta

To ensure the fix addresses the bug, first download the 16.7 beta from TestFlight (any 16.7 build is fine). Log in to the app and before doing anything else, turn off the device wifi and data connections. Then tap the Notifications tab and expect the app to crash.

#### Verify the crash is fixed on this branch

Download the build from this PR and repeat the same steps as above, expect no crash to occur.

#### Regression testing

Mention suggestions should work just as they did before this bug fix was applied. Check mentions in Reader, the Block editor, Notifications replies, comments (My Site > Comments), etc.

Cross-post suggestions – which are available by typing the `+` character (or long-pressing the toolbar button labelled with an @-symbol) in a post on a site that has the O2 plugin installed (e.g. [P2 sites](https://wordpress.com/p2/) – should continue to work as before. These are only available within the Gutenberg editor (not in other places such as comments, replies, etc).

Expect a slight network delay the first time suggestions are displayed for a particular site. However, if they are displayed before one minute has passed, they should be shown immediately.

Suggestions (mentions and cross-posts) should be available when the device is offline, _if_ they were used previously while the device was online.

### PR submission checklist
